# Career Management by Hilton using locations cluster


First we import the libraries, that will be necessary to run the code:

In [1]:
import pandas as pd
!pip install wikitables
from wikitables import import_tables
import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

### Step 1: get the data from Wikitable and process them to get a dataframe

In [2]:
list_hotels = import_tables('List of properties of Hilton Worldwide')
print(list_hotels[0].name)
                

List of properties of Hilton Worldwide[0]


In [3]:
data_hotels= json.loads(list_hotels[0].json())

In [4]:
df= pd.DataFrame(data_hotels)
df

City         Country  \
0                  Tirana         Albania   
1                 Algiers         Algeria   
2            Buenos Aires       Argentina   
3                   Pilar       Argentina   
4                 Yerevan         Armenia   
5              Palm Beach           Aruba   
6                Adelaide       Australia   
7                Brisbane       Australia   
8                  Cairns       Australia   
9                  Darwin       Australia   
10                  Perth       Australia   
11             Gold Coast       Australia   
12             Gold Coast       Australia   
13                 Sydney       Australia   
14                 Vienna         Austria   
15                 Vienna         Austria   
16                 Vienna         Austria   
17                   Baku       Azebaijan   
18                 Nassau     The Bahamas   
19                 Bimini     The Bahamas   
20             Bridgetown        Barbados   
21                  Minsk         Belarus   
22                  Minsk         Belarus   
23                Antwerp         Belgium   
24               Brussels         Belgium   
25               Brussels         Belgium   
26              San Pedro          Belize   
27               Gaborone        Botswana   
28         Belo Horizonte          Brazil   
29                Goiania          Brazil   
...                   ...             ...   
973                Oxford  United Kingdom   
974               Reading  United Kingdom   
975             Sheffield  United Kingdom   
976             Sheffield  United Kingdom   
977             Sheffield  United Kingdom   
978           Southampton  United Kingdom   
979           Southampton  United Kingdom   
980              Stansted  United Kingdom   
981   Stratford-upon-Avon  United Kingdom   
982             Sundeland  United Kingdom   
983               Swindon  United Kingdom   
984         Templepatrick  United Kingdom   
985            Tewkesbury  United Kingdom   
986               Warwick  United Kingdom   
987               Watford  United Kingdom   
988               Wembley  United Kingdom   
989                Woking  United Kingdom   
990             Wokingham  United Kingdom   
991               Woodley  United Kingdom   
992                  York  United Kingdom   
993                  York  United Kingdom   
994                  York  United Kingdom   
995            Montevideo         Uruguay   
996            Montevideo         Uruguay   
997              Tashkent      Uzbekistan   
998              Valencia       Venezuela   
999               Da Nang         Vietnam   
1000                Hanoi         Vietnam   
1001                Hanoi         Vietnam   
1002               Lusaka          Zambia   

                                                   Name        Notes  
0                              Hilton Garden Inn Tirana               
1                                        Hilton Algiers  Coming soon  
2                                   Hilton Buenos Aires               
3                                          Hilton Pilar               
4       DoubleTree by Hilton Hotels Yerevan City Centre               
5                Hilton Aruba Caribbean Resort & Casino               
6                                       Hilton Adelaide               
7                                       Hilton Brisbane               
8                                         Hilton Cairns               
9                                         Hilton Darwin               
10                                Parmelia Hilton Perth               
11                              Hilton Surfers Paradise               
12                    Hilton Surfers Paradise Residence               
13                                        Hilton Sydney               
14                                        Hilton Vienna               
15                      Hilton Vienna Danube Waterfront               
16                     

In [5]:
df= df.drop(['Notes'], axis=1)

### Step 2: get the list of European countries
As we want to limit the study to european countries, we first get a list of the corresponding countries.    <br>
The list is from Wikipedia as well. The datas a processed the same way as before.

In [6]:
list_countries = import_tables('List of European countries by area')
print(list_countries[0].name)

List of European countries by area[0]


In [7]:
data_countries= json.loads(list_countries[0].json())
df_countries= pd.DataFrame(data_countries)
df_countries

Notes  Rank  \
0      17,098,242 &nbsp; km 2 including Northern Asia     1   
1                                                         2   
2   643,801 &nbsp; km 2 when the overseas departme...     3   
3   505,990 &nbsp; km 2 when the Canary Islands , ...     4   
4                                                         5   
5                This includes Svalbard and Jan Mayen     6   
6                                                         7   
7                                                         8   
8                                                         9   
9                                                        10   
10  260,665 &nbsp; km 2 when the British Overseas ...    11   
11                                                       12   
12                                                       13   
13         2,724,902 &nbsp; km 2 including Asian part    14   
14                                                       15   
15                                                       16   
16                                                       17   
17                                                       18   
18                                                       19   
19                                                       20   
20                                                       21   
21                88,361 &nbsp; km 2 including Kosovo    22   
22                                                       23   
23                                                       24   
24                                                       25   
25                                                       26   
26                                                       27   
27                                                       28   
28                                                       29   
29  This includes Faroe Islands ; 2,210,579 &nbsp;...    30   
30                                                       31   
31  Excluding Caribbean Netherlands , Aruba , Cura...    32   
32                                                       33   
33                                                       34   
34                  Sometimes considered part of Asia    35   
35                                                       36   
36                                                       37   
37           783,562 &nbsp; km 2 including Asian part    38   
38                                                       39   
39                                                       40   
40                         Partially recognised state    41   
41                  Sometimes considered part of Asia    42   
42            86,600 &nbsp; km 2 including Asian part    43   
43                                                       44   
44            69,700 &nbsp; km 2 including Asian part    45   
45                                                       46   
46                                                       47   
47                                                       48   
48                                                       49   
49                                                       50   
50                                                       51   

                     State Total area (km2)  
0                 Russia *          3972400  
1                  Ukraine           603628  
2                 France *           551695  
3                  Spain *           498511  
4                   Sweden           450295  
5                   Norway           385178  
6                  Germany           357386  
7                  Finland           338145  
8                   Poland           312685  
9                    Italy           301338  
10        United Kingdom *           242495  
11                 Romania           238397  
12                 Belarus           207600  
13            Kazakhstan *   180,000 (est.)  
14                  Greece           131940  
15                Bulgaria         

In [8]:
df_countries = df_countries.drop(['Notes', 'Rank', 'Total area (km2)'], axis=1)
df_countries

State
0                 Russia *
1                  Ukraine
2                 France *
3                  Spain *
4                   Sweden
5                   Norway
6                  Germany
7                  Finland
8                   Poland
9                    Italy
10        United Kingdom *
11                 Romania
12                 Belarus
13            Kazakhstan *
14                  Greece
15                Bulgaria
16                 Iceland
17                 Hungary
18                Portugal
19                 Austria
20          Czech Republic
21                  Serbia
22                 Ireland
23               Lithuania
24                  Latvia
25                 Croatia
26  Bosnia and Herzegovina
27                Slovakia
28                 Estonia
29               Denmark *
30             Switzerland
31           Netherlands *
32                 Moldova
33                 Belgium
34               Armenia *
35                 Albania
36         North Macedonia
37                Turkey *
38                Slovenia
39              Montenegro
40                  Kosovo
41                  Cyprus
42            Azerbaijan *
43              Luxembourg
44               Georgia *
45                 Andorra
46                   Malta
47           Liechtenstein
48              San Marino
49                  Monaco
50            Vatican City

Let's clean up the names in order to have correspondance to the former data frame.

In [9]:
for i in range(len(df_countries)):
    if df_countries['State'][i][-1] == '*':
        df_countries['State'][i] = df_countries['State'][i][:-2]

In [10]:
df_countries

State
0                   Russia
1                  Ukraine
2                   France
3                    Spain
4                   Sweden
5                   Norway
6                  Germany
7                  Finland
8                   Poland
9                    Italy
10          United Kingdom
11                 Romania
12                 Belarus
13              Kazakhstan
14                  Greece
15                Bulgaria
16                 Iceland
17                 Hungary
18                Portugal
19                 Austria
20          Czech Republic
21                  Serbia
22                 Ireland
23               Lithuania
24                  Latvia
25                 Croatia
26  Bosnia and Herzegovina
27                Slovakia
28                 Estonia
29                 Denmark
30             Switzerland
31             Netherlands
32                 Moldova
33                 Belgium
34                 Armenia
35                 Albania
36         North Macedonia
37                  Turkey
38                Slovenia
39              Montenegro
40                  Kosovo
41                  Cyprus
42              Azerbaijan
43              Luxembourg
44                 Georgia
45                 Andorra
46                   Malta
47           Liechtenstein
48              San Marino
49                  Monaco
50            Vatican City

### Step 3: Filter the locations and make them unique

In [11]:
df_europe = pd.DataFrame(columns=df.columns)
df_europe

Empty DataFrame
Columns: [City, Country, Name]
Index: []

In [12]:
for i in range(len(df)):
    if df['Country'][i] in df_countries['State'].values:
        df_europe=df_europe.append(df.loc[i], ignore_index=True)
df_europe

City         Country  \
0                 Tirana         Albania   
1                Yerevan         Armenia   
2                 Vienna         Austria   
3                 Vienna         Austria   
4                 Vienna         Austria   
5                  Minsk         Belarus   
6                  Minsk         Belarus   
7                Antwerp         Belgium   
8               Brussels         Belgium   
9               Brussels         Belgium   
10                 Sofia        Bulgaria   
11    Varna/Golden Sands        Bulgaria   
12             Dubrovnik         Croatia   
13                Zagreb         Croatia   
14                Zagreb         Croatia   
15               Nicosia          Cyprus   
16                Prague  Czech Republic   
17                Prague  Czech Republic   
18               Tallinn         Estonia   
19              Helsinki         Finland   
20              Helsinki         Finland   
21                Vantaa         Finland   
22       Evian Les Bains          France   
23      La Seyne-Sur-Mer          France   
24                 Paris          France   
25                 Paris          France   
26                 Paris          France   
27                 Paris          France   
28            Roissypole          France   
29            Strasbourg          France   
..                   ...             ...   
377              Newbury  United Kingdom   
378            Newcastle  United Kingdom   
379            Newcastle  United Kingdom   
380              Newport  United Kingdom   
381    North Queensferry  United Kingdom   
382          Northampton  United Kingdom   
383           Nottingham  United Kingdom   
384           Nottingham  United Kingdom   
385               Oxford  United Kingdom   
386              Reading  United Kingdom   
387            Sheffield  United Kingdom   
388            Sheffield  United Kingdom   
389            Sheffield  United Kingdom   
390          Southampton  United Kingdom   
391          Southampton  United Kingdom   
392             Stansted  United Kingdom   
393  Stratford-upon-Avon  United Kingdom   
394            Sundeland  United Kingdom   
395              Swindon  United Kingdom   
396        Templepatrick  United Kingdom   
397           Tewkesbury  United Kingdom   
398              Warwick  United Kingdom   
399              Watford  United Kingdom   
400              Wembley  United Kingdom   
401               Woking  United Kingdom   
402            Wokingham  United Kingdom   
403              Woodley  United Kingdom   
404                 York  United Kingdom   
405                 York  United Kingdom   
406                 York  United Kingdom   

                                                  Name  
0                             Hilton Garden Inn Tirana  
1      DoubleTree by Hilton Hotels Yerevan City Centre  
2                                        Hilton Vienna  
3                      Hilton Vienna Danube Waterfront  
4                                  Hilton Vienna Plaza  
5                     DoubleTree by Hilton Hotel Minsk  
6                  Hampton by Hilton Minsk City Centre  
7                           Hilton Antwerp City Centre  
8                                 Hilton Brussels City  
9                          Hilton Brussels Grand Place  
10                                        Hilton Sofia  
11     DoubleTree by Hilton Hotel Varna - Golden Sands  
12                           Hilton Imperial Dubrovnik  
13               Canopy by Hilton Zagreb - City Centre  
14                   DoubleTree by Hilton Hotel Zagreb  
15                                      Hilton Nicosia  
16                                       Hilton Prague  
17                              Hilton Prague Old Town  
18                                 Hilton Tallinn Park  
19                     Hilton Helsinki Kalastajatorppa  
20                              Hilton Helsinki Strand  
21                             Hilton

In [13]:
distinct_locations_europe = df_europe.groupby('City')['Name'].count().to_frame()
distinct_locations_europe.head()

Name
City          
Aachen       1
Aberdeen     4
Adana        1
Adiyaman     1
Alamaty      1

In [14]:
distinct_locations_europe.reset_index(inplace = True)
distinct_locations_europe.columns= ['City', 'Qty_hotels']
distinct_locations_europe

City  Qty_hotels
0                Aachen           1
1              Aberdeen           4
2                 Adana           1
3              Adiyaman           1
4               Alamaty           1
5               Algarve           1
6             Amsterdam           6
7                Ankara           2
8               Antakya           1
9               Antalya           1
10              Antwerp           1
11              Arundel           1
12               Astana           3
13               Athens           1
14               Avanos           1
15             Aviemore           1
16               Balice           1
17             Ballater           1
18            Barcelona           3
19           Basaksehir           1
20                 Bath           1
21             Bayrakli           1
22              Belfast           2
23             Belgrade           1
24            Bellshill           1
25               Berlin           5
26           Birmingham           5
27            Blackpool           1
28                 Bolu           1
29                 Bonn           1
..                  ...         ...
201           The Hague           1
202              Tirana           1
203             Trabzon           1
204               Turin           1
205              Tyumen           1
206                 Ufa           2
207           Ulyanovsk           1
208            Valletta           1
209                 Van           1
210              Vantaa           1
211  Varna/Golden Sands           1
212              Venice           2
213          Versailles           1
214              Vienna           3
215           Vilamoura           1
216            Volgorad           2
217            Voronezh           1
218              Warsaw           5
219             Warwick           1
220             Watford           1
221             Wembley           1
222              Woking           1
223           Wokingham           1
224             Woodley           1
225             Wroclaw           1
226              Yalova           1
227             Yerevan           1
228                York           3
229              Zagreb           2
230              Zurich           2

[231 rows x 2 columns]

### Step 4: use Geopy and add Latitude and Longitude to each city

Geopy limits somehow the loops to 17 rows. Let's run the request several times to get the 231 answers.

In [171]:
!pip install geopy
from geopy.geocoders import Nominatim 

In [299]:
address = distinct_locations_europe['City'][150] # Aachen
geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
distinct_locations_europe.loc[150, 'Latitude'] = location.latitude
distinct_locations_europe.loc[150, 'Longitude'] = location.longitude
distinct_locations_europe.head()

City  Qty_hotels   Latitude  Longitude
0    Aachen           2  50.776351   6.083862
1  Aberdeen           8  57.148243  -2.092809
2     Adana           2  37.143800  35.498409
3  Adiyaman           2  37.789360  38.314110
4   Alamaty           2  43.238949  76.889709

In [244]:
for i in range(len(distinct_locations_europe)):
    address = distinct_locations_europe['City'][i]
    geolocator = Nominatim(user_agent="foursquare_agent")
    location = geolocator.geocode(address)
    if location != None:
        distinct_locations_europe.loc[i, 'Latitude'] = location.latitude
        distinct_locations_europe.loc[i, 'Longitude'] = location.longitude
        
distinct_locations_europe

KeyError: 231

In [245]:
distinct_locations_europe

City  Qty_hotels   Latitude   Longitude
0                Aachen           2  50.776351    6.083862
1              Aberdeen           8  57.148243   -2.092809
2                 Adana           2  37.143800   35.498409
3              Adiyaman           2  37.789360   38.314110
4               Alamaty           2        NaN         NaN
5               Algarve           2  37.245425   -8.150925
6             Amsterdam          12  52.372760    4.893604
7                Ankara           4  39.920777   32.854067
8               Antakya           2  36.219114   36.161628
9               Antalya           2  36.927965   30.727687
10              Antwerp           2  51.221110    4.399708
11              Arundel           2  50.854538   -0.564816
12               Astana           6  51.150921   71.438860
13               Athens           2  37.983941   23.728305
14               Avanos           2  38.865858   34.850364
15             Aviemore           2  57.193753   -3.828750
16               Balice           2  50.086830   19.792135
17             Ballater           2  57.049506   -3.040517
18            Barcelona           6  41.382894    2.177432
19           Basaksehir           2  41.097693   28.806163
20                 Bath           2  51.381386   -2.359696
21             Bayrakli           2  38.486739   27.149265
22              Belfast           4  54.596441   -5.930276
23             Belgrade           2  44.817813   20.456897
24            Bellshill           2  55.818931   -4.022579
25               Berlin          10  52.517037   13.388860
26           Birmingham          10  52.479699   -1.902691
27            Blackpool           2  53.824548   -3.037320
28                 Bolu           2  40.621210   31.646026
29                 Bonn           2  50.735851    7.100660
..                  ...         ...        ...         ...
201           The Hague           2  52.079984    4.311346
202              Tirana           2  41.315886   19.900912
203             Trabzon           2  40.884986   39.822206
204               Turin           2  45.067755    7.682489
205              Tyumen           2  57.153534   65.542274
206                 Ufa           4  54.726288   55.947727
207           Ulyanovsk           2  54.310759   48.364277
208            Valletta           2  35.898982   14.513676
209                 Van           2  13.290403  108.426511
210              Vantaa           2  60.309187   25.036453
211  Varna/Golden Sands           2  43.285475   28.044049
212              Venice           4  45.437191   12.334590
213          Versailles           2  48.803540    2.126689
214              Vienna           6  48.208354   16.372504
215           Vilamoura           2  37.075950   -8.116540
216            Volgorad           4        NaN         NaN
217            Voronezh           2  51.660598   39.200586
218              Warsaw          10  52.233717   21.071411
219             Warwick           2  41.700202  -71.416111
220             Watford           2  51.655388   -0.395742
221             Wembley           2  51.553155   -0.295756
222              Woking           2  51.320189   -0.556473
223           Wokingham           2  51.457085   -0.886537
224             Woodley           2  51.453044   -0.904690
225             Wroclaw           2  51.126311   16.978196
226              Yalova           2  40.629185   29.250026
227             Yerevan           2  40.177612   44.512585
228                York           6  53.959055   -1.081536
229              Zagreb           4  45.813177   15.977048
230              Zurich           4  47.372394    8.542333

[231 rows x 4 columns]

Let's check the missing values and correct them manually.

In [258]:
distinct_locations_europe.sort_values(['Latitude']).tail(15)

City  Qty_hotels   Latitude  Longitude
15           Aviemore           2  57.193753  -3.828750
105             Kirov           2  58.603526  49.663903
163              Perm           2  58.884391  56.438995
190         Stockholm           2  59.325117  18.071093
197           Tallinn           2  59.437216  24.745369
174  Saint Petersburg           4  59.960674  30.158655
91           Helsinki           4  60.167410  24.942577
210            Vantaa           2  60.309187  25.036453
168         Reykjavik           6  64.145981 -21.942237
111       Krasnoyarsk           2  66.624407  94.742863
4             Alamaty           2        NaN        NaN
147           Newport           2        NaN        NaN
154        Novosbirsk           2        NaN        NaN
194         Sundeland           2        NaN        NaN
216          Volgorad           4        NaN        NaN

In [15]:
#Fill up the missing values manually with datas from https://www.latlong.net

distinct_locations_europe.loc[4, 'Latitude'] = 43.238949
distinct_locations_europe.loc[4, 'Longitude'] = 76.889709

distinct_locations_europe.loc[147, 'Latitude'] = 51.58774
distinct_locations_europe.loc[147, 'Longitude'] = -2.99835

distinct_locations_europe.loc[194, 'Latitude'] = 54.906101
distinct_locations_europe.loc[194, 'Longitude'] = -1.381130

distinct_locations_europe.loc[216, 'Latitude'] = 48.700001
distinct_locations_europe.loc[216, 'Longitude'] = 44.516666


Make a copy of the Data Frame as csv file

In [293]:
from project_lib import Project
project = Project(None, "....","...")

In [294]:
project.save_data(file_name = "Hilton.csv",data = distinct_locations_europe.to_csv(index=False))

{'file_name': 'Hilton.csv',
 'message': 'File saved to project storage.',
 'bucket_name': 'courseracapstone-donotdelete-pr-iwfz5ysqserla3',
 'asset_id': '34571736-df8f-4c9b-80d2-3336b740a20e'}

In [16]:
# The code was removed by Watson Studio for sharing.

City  Qty_hotels   Latitude  Longitude
0    Aachen           2  37.245425  -8.150925
1  Aberdeen           8  57.148243  -2.092809
2     Adana           2  37.143800  35.498409
3  Adiyaman           2  37.789360  38.314110
4   Alamaty           2  43.238949  76.889709

Novosibirsk is in Russia, but not in Europe. We can drop the missing row

In [17]:
distinct_locations_europe = distinct_locations_europe.drop([154])

In [18]:
distinct_locations_europe.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 230 entries, 0 to 230
Data columns (total 4 columns):
City          230 non-null object
Qty_hotels    230 non-null int64
Latitude      230 non-null float64
Longitude     230 non-null float64
dtypes: float64(2), int64(1), object(1)
memory usage: 9.0+ KB


After running the complete code, some errors were found on the map. Some cities have the same name. Let's correct the errors manually:

In [19]:
distinct_locations_europe[(distinct_locations_europe['City']=='Siena')|(distinct_locations_europe['City']=='Van')|(distinct_locations_europe['City']=='Sofia')|(distinct_locations_europe['City']=='Humberside')|(distinct_locations_europe['City']=='Warwick')|(distinct_locations_europe['City']=='Lincoln')]

City  Qty_hotels   Latitude   Longitude
94   Humberside           2  43.657694  -79.492512
119     Lincoln           2  47.829907 -118.414727
182       Siena           2  35.000074  104.999927
185       Sofia           2 -15.253840   48.256216
209         Van           2  13.290403  108.426511
219     Warwick           2  41.700202  -71.416111

In [20]:
#Fill up the missing values manually with datas from https://www.latlong.net

distinct_locations_europe.loc[94, 'Latitude'] = 53.747372
distinct_locations_europe.loc[94, 'Longitude'] = -0.338653

distinct_locations_europe.loc[119, 'Latitude'] = 53.234444
distinct_locations_europe.loc[119, 'Longitude'] = -0.538611

distinct_locations_europe.loc[182, 'Latitude'] = 43.318611
distinct_locations_europe.loc[182, 'Longitude'] = 11.330556

distinct_locations_europe.loc[185, 'Latitude'] = 42.698334
distinct_locations_europe.loc[185, 'Longitude'] = 23.319941

distinct_locations_europe.loc[209, 'Latitude'] = 38.499817
distinct_locations_europe.loc[209, 'Longitude'] = 43.378143

distinct_locations_europe.loc[219, 'Latitude'] = 52.2833
distinct_locations_europe.loc[219, 'Longitude'] = -1.5833

distinct_locations_europe.loc[0, 'Latitude'] = 50.775555
distinct_locations_europe.loc[0, 'Longitude'] = 6.083611

In [21]:
df = distinct_locations_europe
df

City  Qty_hotels   Latitude  Longitude
0                Aachen           2  50.775555   6.083611
1              Aberdeen           8  57.148243  -2.092809
2                 Adana           2  37.143800  35.498409
3              Adiyaman           2  37.789360  38.314110
4               Alamaty           2  43.238949  76.889709
5               Algarve           2  37.245425  -8.150925
6             Amsterdam          12  52.372760   4.893604
7                Ankara           4  39.920777  32.854067
8               Antakya           2  36.219114  36.161628
9               Antalya           2  36.927965  30.727687
10              Antwerp           2  51.221110   4.399708
11              Arundel           2  50.854538  -0.564816
12               Astana           6  51.150921  71.438860
13               Athens           2  37.983941  23.728305
14               Avanos           2  38.865858  34.850364
15             Aviemore           2  57.193753  -3.828750
16               Balice           2  50.086830  19.792135
17             Ballater           2  57.049506  -3.040517
18            Barcelona           6  41.382894   2.177432
19           Basaksehir           2  41.097693  28.806163
20                 Bath           2  51.381386  -2.359696
21             Bayrakli           2  38.486739  27.149265
22              Belfast           4  54.596441  -5.930276
23             Belgrade           2  44.817813  20.456897
24            Bellshill           2  55.818931  -4.022579
25               Berlin          10  52.517036  13.388860
26           Birmingham          10  52.479699  -1.902691
27            Blackpool           2  53.824548  -3.037320
28                 Bolu           2  40.621210  31.646026
29                 Bonn           2  50.735851   7.100660
..                  ...         ...        ...        ...
201           The Hague           2  52.079984   4.311346
202              Tirana           2  41.315886  19.900912
203             Trabzon           2  40.884986  39.822206
204               Turin           2  45.067755   7.682489
205              Tyumen           2  57.153534  65.542274
206                 Ufa           4  54.726288  55.947727
207           Ulyanovsk           2  54.310759  48.364277
208            Valletta           2  35.898982  14.513676
209                 Van           2  38.499817  43.378143
210              Vantaa           2  60.309187  25.036453
211  Varna/Golden Sands           2  43.285475  28.044049
212              Venice           4  45.437191  12.334590
213          Versailles           2  48.803540   2.126689
214              Vienna           6  48.208354  16.372504
215           Vilamoura           2  37.075950  -8.116540
216            Volgorad           4  48.700001  44.516666
217            Voronezh           2  51.660598  39.200586
218              Warsaw          10  52.233717  21.071411
219             Warwick           2  52.283300  -1.583300
220             Watford           2  51.655388  -0.395742
221             Wembley           2  51.553155  -0.295756
222              Woking           2  51.320189  -0.556473
223           Wokingham           2  51.457085  -0.886537
224             Woodley           2  51.453044  -0.904690
225             Wroclaw           2  51.126311  16.978196
226              Yalova           2  40.629185  29.250026
227             Yerevan           2  40.177612  44.512585
228                York           6  53.959056  -1.081536
229              Zagreb           4  45.813177  15.977048
230              Zurich           4  47.372394   8.542333

[230 rows x 4 columns]

### Step 5: run the cluster using k-Means

In [22]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!pip install folium==0.5.0 
#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [23]:
#Define Foursquare

CLIENT_ID = '....' # your Foursquare ID
CLIENT_SECRET = '2X32A5BCJQPI1C5BFS5FHSWG0KIHMNEIJDHGYV0IHXHTFYCV' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: F0CC5OWAXLXZHWAUYGZR3PT4QA4H3TH5CGWYULIY2B4BNKYJ
CLIENT_SECRET:2X32A5BCJQPI1C5BFS5FHSWG0KIHMNEIJDHGYV0IHXHTFYCV


In [24]:
df.loc[0, 'City']
city_latitude = df.loc[0, 'Latitude'] # city latitude value
city_longitude = df.loc[0, 'Longitude'] # city longitude value

city_name = df.loc[0, 'City'] # city name

print('Latitude and longitude values of {} are {}, {}.'.format(city_name, 
                                                              city_latitude, 
                                                               city_longitude))

Latitude and longitude values of Aachen are 50.775555, 6.083611.


In [25]:
LIMIT = 200 # limit of number of venues returned by Foursquare API
radius = 600 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    city_latitude, 
    city_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=F0CC5OWAXLXZHWAUYGZR3PT4QA4H3TH5CGWYULIY2B4BNKYJ&client_secret=2X32A5BCJQPI1C5BFS5FHSWG0KIHMNEIJDHGYV0IHXHTFYCV&v=20180605&ll=50.775555,6.083611&radius=600&limit=200'

In [26]:
results = requests.get(url).json()

In [27]:
results

{'meta': {'code': 429,
  'errorType': 'quota_exceeded',
  'errorDetail': 'Quota exceeded',
  'requestId': '5e7e0517dd0f850028d6e5e2'},
 'response': {}}

In [139]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [149]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name          categories        lat       lng
0  Aachener Dom St. Marien              Church  50.774702  6.084103
1                Katschhof               Plaza  50.775611  6.083909
2                   Domhof               Plaza  50.774658  6.083297
3                      AKL  Falafel Restaurant  50.776967  6.083277
4         Vertical Weinbar            Wine Bar  50.776288  6.081469

In [150]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

78 venues were returned by Foursquare.


In [151]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [153]:
hilton_venues = getNearbyVenues(names=df['City'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

Aachen


KeyError: 'groups'

In [144]:
print(hilton_venues.shape)
hilton_venues.head()

(10416, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0       Aachen              50.775555                6.083611   
1       Aachen              50.775555                6.083611   
2       Aachen              50.775555                6.083611   
3       Aachen              50.775555                6.083611   
4       Aachen              50.775555                6.083611   

                     Venue  Venue Latitude  Venue Longitude  \
0  Aachener Dom St. Marien       50.774702         6.084103   
1                Katschhof       50.775611         6.083909   
2                   Domhof       50.774658         6.083297   
3                      AKL       50.776967         6.083277   
4         Vertical Weinbar       50.776288         6.081469   

       Venue Category  
0              Church  
1               Plaza  
2               Plaza  
3  Falafel Restaurant  
4            Wine Bar

In [82]:
 # Make a copy of the Data Frame as csv file

from project_lib import Project
project = Project(None, "cc243e6e-b797-456c-945c-156363111c72","p-8a5d93279d04cef93ae83211150ff54e2cec2838")

project.save_data(file_name = "Hilton_venues.csv",data = hilton_venues.to_csv(index=False), overwrite = True)

{'file_name': 'Hilton_venues.csv',
 'message': 'File saved to project storage.',
 'bucket_name': 'courseracapstone-donotdelete-pr-iwfz5ysqserla3',
 'asset_id': 'cad434ff-f793-443f-9240-307580c99e8a'}

In [28]:
#... so you can start again working on the next day even if Foursquare's quota is reached
body = client_ca49a1412575485390e9c38bb8931ea4.get_object(Bucket='courseracapstone-donotdelete-pr-iwfz5ysqserla3',Key='Hilton_venues.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

hilton_venues = pd.read_csv(body)
hilton_venues.head()


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0       Aachen              50.775555                6.083611   
1       Aachen              50.775555                6.083611   
2       Aachen              50.775555                6.083611   
3       Aachen              50.775555                6.083611   
4       Aachen              50.775555                6.083611   

                     Venue  Venue Latitude  Venue Longitude  \
0  Aachener Dom St. Marien       50.774702         6.084103   
1                Katschhof       50.775611         6.083909   
2                   Domhof       50.774658         6.083297   
3                      AKL       50.776967         6.083277   
4         Vertical Weinbar       50.776288         6.081469   

       Venue Category  
0              Church  
1               Plaza  
2               Plaza  
3  Falafel Restaurant  
4            Wine Bar

##### Analyze each Neighborhood

In [29]:
# one hot encoding
hilton_onehot = pd.get_dummies(hilton_venues[['Venue Category']], prefix="", prefix_sep="")

# add City column back to dataframe
hilton_onehot['Neighborhood'] = hilton_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [hilton_onehot.columns[-1]] + list(hilton_onehot.columns[:-1])
hilton_onehot = hilton_onehot[fixed_columns]

hilton_onehot.head()

Çöp Şiş Place  ATM  Accessories Store  African Restaurant  Airport  \
0              0    0                  0                   0        0   
1              0    0                  0                   0        0   
2              0    0                  0                   0        0   
3              0    0                  0                   0        0   
4              0    0                  0                   0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   Alsatian Restaurant  ...  Waterfront  Whisky Bar  Wine Bar  Wine Shop  \
0                    0  ...           0           0         0          0   
1                    0  ...           0           0         0          0   
2                    0  ...           0           0         0          0   
3                    0  ...           0           0         0          0   
4                    0  ...           0           0         1          0   

   Winery  Wings Joint  Women's Store  Yoga Studio  Zoo  Zoo Exhibit  
0       0            0              0            0    0            0  
1       0            0              0            0    0            0  
2       0            0              0            0    0            0  
3       0            0              0            0    0            0  
4       0            0              0            0    0            0  

[5 rows x 449 columns]

#### .. and group them by occurence of each category

In [30]:
hilton_grouped = hilton_onehot.groupby('Neighborhood').mean().reset_index()
hilton_grouped

Neighborhood  Çöp Şiş Place  ATM  Accessories Store  \
0                Aachen       0.000000  0.0                0.0   
1              Aberdeen       0.000000  0.0                0.0   
2                 Adana       0.000000  0.0                0.0   
3               Alamaty       0.000000  0.0                0.0   
4             Amsterdam       0.000000  0.0                0.0   
5                Ankara       0.000000  0.0                0.0   
6               Antakya       0.000000  0.0                0.0   
7               Antalya       0.000000  0.0                0.0   
8               Antwerp       0.000000  0.0                0.0   
9               Arundel       0.000000  0.0                0.0   
10               Astana       0.000000  0.0                0.0   
11               Athens       0.000000  0.0                0.0   
12             Aviemore       0.000000  0.0                0.0   
13               Balice       0.000000  0.0                0.0   
14             Ballater       0.000000  0.0                0.0   
15            Barcelona       0.000000  0.0                0.0   
16           Basaksehir       0.035714  0.0                0.0   
17                 Bath       0.000000  0.0                0.0   
18             Bayrakli       0.000000  0.0                0.0   
19              Belfast       0.000000  0.0                0.0   
20             Belgrade       0.000000  0.0                0.0   
21            Bellshill       0.000000  0.0                0.0   
22               Berlin       0.000000  0.0                0.0   
23           Birmingham       0.000000  0.0                0.0   
24            Blackpool       0.000000  0.0                0.0   
25                 Bolu       0.000000  0.0                0.0   
26                 Bonn       0.000000  0.0                0.0   
27           Bottesford       0.000000  0.0                0.0   
28          Bournemouth       0.000000  0.0                0.0   
29            Bracknell       0.000000  0.0                0.0   
..                  ...            ...  ...                ...   
170             Tallinn       0.000000  0.0                0.0   
171          Tewkesbury       0.000000  0.0                0.0   
172           The Hague       0.000000  0.0                0.0   
173               Turin       0.000000  0.0                0.0   
174              Tyumen       0.000000  0.0                0.0   
175                 Ufa       0.000000  0.0                0.0   
176           Ulyanovsk       0.000000  0.0                0.0   
177            Valletta       0.000000  0.0                0.0   
178                 Van       0.000000  0.0                0.0   
179              Vantaa       0.000000  0.0                0.0   
180  Varna/Golden Sands       0.000000  0.0                0.0   
181              Venice       0.000000  0.0                0.0   
182          Versailles       0.000000  0.0                0.0   
183              Vienna       0.000000  0.0                0.0   
184           Vilamoura       0.000000  0.0                0.0   
185            Volgorad       0.000000  0.0                0.0   
186            Voronezh       0.000000  0.0                0.0   
187              Warsaw       0.000000  0.0                0.0   
188             Warwick       0.000000  0.0                0.0   
189             Watford       0.000000  0.0                0.0   
190             Wembley       0.000000  0.0                0.0   
191              Woking       0.000000  0.0                0.0   
192           Wokingham       0.000000  0.0                0.0   
193             Woodley       0.000000  0.0                0.0   
194             Wroclaw       0.000000  0.0                0.0   
195              Yalova       0.000000  0.0                0.0   
196             Yerevan       0.000000  0.0                0.0   
197                York       0.000000  0.0                0.0   
198              Zagreb       0.000000  0.0        

Make a dataframe and process it:

In [31]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [32]:
import numpy as np

In [33]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
hilton_venues_sorted = pd.DataFrame(columns=columns)
hilton_venues_sorted['Neighborhood'] = hilton_grouped['Neighborhood']

for ind in np.arange(hilton_grouped.shape[0]):
    hilton_venues_sorted.iloc[ind, 1:] = return_most_common_venues(hilton_grouped.iloc[ind, :], num_top_venues)

hilton_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0       Aachen             Drugstore                  Café   
1     Aberdeen              Beer Bar                   Bar   
2        Adana         Moving Target           Zoo Exhibit   
3      Alamaty      Asian Restaurant                  Café   
4    Amsterdam                 Hotel                   Bar   

  3rd Most Common Venue          4th Most Common Venue 5th Most Common Venue  \
0        Ice Cream Shop                         Bakery    Italian Restaurant   
1    Seafood Restaurant          Performing Arts Venue        Sandwich Place   
2           Flea Market                    Fabric Shop               Factory   
3           Karaoke Bar  Paper / Office Supplies Store    Turkish Restaurant   
4           Coffee Shop           Marijuana Dispensary        Clothing Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                 Plaza                   Bar           Coffee Shop   
1           Supermarket           Pizza Place                Museum   
2    Falafel Restaurant                  Farm        Farmers Market   
3              Beer Bar         Grocery Store                   Gym   
4                  Café     French Restaurant          Cocktail Bar   

  9th Most Common Venue 10th Most Common Venue  
0           Snack Place      German Restaurant  
1             Multiplex          Movie Theater  
2  Fast Food Restaurant                  Field  
3           Sports Club                    Pub  
4    Italian Restaurant                 Bakery

In [34]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [37]:
# set number of clusters
kclusters = 6

hilton_grouped_clustering = hilton_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, n_init=12, random_state=0).fit(hilton_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 2, 0, 0, 0, 0, 0, 0, 3], dtype=int32)

In [36]:
# Reset the dataframe for more trials
hilton_venues_sorted = pd.DataFrame(columns=columns)
hilton_venues_sorted['Neighborhood'] = hilton_grouped['Neighborhood']

for ind in np.arange(hilton_grouped.shape[0]):
    hilton_venues_sorted.iloc[ind, 1:] = return_most_common_venues(hilton_grouped.iloc[ind, :], num_top_venues)

hilton_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0       Aachen             Drugstore                  Café   
1     Aberdeen              Beer Bar                   Bar   
2        Adana         Moving Target           Zoo Exhibit   
3      Alamaty      Asian Restaurant                  Café   
4    Amsterdam                 Hotel                   Bar   

  3rd Most Common Venue          4th Most Common Venue 5th Most Common Venue  \
0        Ice Cream Shop                         Bakery    Italian Restaurant   
1    Seafood Restaurant          Performing Arts Venue        Sandwich Place   
2           Flea Market                    Fabric Shop               Factory   
3           Karaoke Bar  Paper / Office Supplies Store    Turkish Restaurant   
4           Coffee Shop           Marijuana Dispensary        Clothing Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                 Plaza                   Bar           Coffee Shop   
1           Supermarket           Pizza Place                Museum   
2    Falafel Restaurant                  Farm        Farmers Market   
3              Beer Bar         Grocery Store                   Gym   
4                  Café     French Restaurant          Cocktail Bar   

  9th Most Common Venue 10th Most Common Venue  
0           Snack Place      German Restaurant  
1             Multiplex          Movie Theater  
2  Fast Food Restaurant                  Field  
3           Sports Club                    Pub  
4    Italian Restaurant                 Bakery

In [38]:
# add clustering labels
hilton_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

hilton_merged = df

# merge hilton_grouped with df to add latitude/longitude for each neighborhood
hilton_merged = hilton_merged.join(hilton_venues_sorted.set_index('Neighborhood'), on='City')

hilton_merged.head() # check the last columns!

City  Qty_hotels   Latitude  Longitude  Cluster Labels  \
0    Aachen           2  50.775555   6.083611             0.0   
1  Aberdeen           8  57.148243  -2.092809             0.0   
2     Adana           2  37.143800  35.498409             2.0   
3  Adiyaman           2  37.789360  38.314110             NaN   
4   Alamaty           2  43.238949  76.889709             0.0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0             Drugstore                  Café        Ice Cream Shop   
1              Beer Bar                   Bar    Seafood Restaurant   
2         Moving Target           Zoo Exhibit           Flea Market   
3                   NaN                   NaN                   NaN   
4      Asian Restaurant                  Café           Karaoke Bar   

           4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                         Bakery    Italian Restaurant                 Plaza   
1          Performing Arts Venue        Sandwich Place           Supermarket   
2                    Fabric Shop               Factory    Falafel Restaurant   
3                            NaN                   NaN                   NaN   
4  Paper / Office Supplies Store    Turkish Restaurant              Beer Bar   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                   Bar           Coffee Shop           Snack Place   
1           Pizza Place                Museum             Multiplex   
2                  Farm        Farmers Market  Fast Food Restaurant   
3                   NaN                   NaN                   NaN   
4         Grocery Store                   Gym           Sports Club   

  10th Most Common Venue  
0      German Restaurant  
1          Movie Theater  
2                  Field  
3                    NaN  
4                    Pub

In [39]:
# Localize the center of the future map in Vienna
center_lat = distinct_locations_europe.loc[214, 'Latitude']
center_long = distinct_locations_europe.loc[214, 'Longitude']

In [40]:
hilton_merged['Cluster Labels'].value_counts()

0.0    127
3.0     68
4.0      2
5.0      1
1.0      1
2.0      1
Name: Cluster Labels, dtype: int64

In [41]:
hilton_merged['Cluster Labels'].values

array([ 0.,  0.,  2., nan,  0., nan,  0.,  0.,  0.,  0.,  0.,  3.,  0.,
        0., nan,  0.,  3.,  3.,  0.,  0.,  3.,  3.,  0.,  0.,  3.,  0.,
        0.,  3.,  4.,  0.,  3.,  0.,  3.,  0.,  0.,  0.,  0.,  0.,  3.,
        0.,  3.,  3., nan,  0.,  0.,  3.,  0.,  0.,  3.,  3.,  0., nan,
        3.,  3.,  3.,  3.,  0., nan,  3.,  3.,  0.,  0.,  3.,  3.,  3.,
        0.,  0.,  3.,  0.,  0.,  1., nan, nan,  0.,  0.,  3.,  0.,  0.,
        0.,  0.,  3.,  3.,  5.,  3.,  3.,  0.,  0.,  3.,  0.,  3.,  3.,
        0.,  0.,  0.,  3.,  0., nan,  3., nan,  0.,  0., nan,  0.,  4.,
        0.,  0.,  0.,  0.,  0.,  3.,  0., nan,  0., nan,  0.,  0., nan,
        0.,  0.,  0.,  0.,  3.,  0.,  0.,  0.,  3.,  3.,  3.,  0.,  3.,
        0.,  0., nan,  0.,  0., nan,  0.,  3.,  0.,  0.,  3., nan, nan,
        0.,  0.,  3.,  0.,  3.,  0.,  0.,  3.,  0.,  0.,  0.,  0., nan,
        0., nan,  0.,  0.,  0.,  0., nan,  0.,  0.,  0.,  3.,  0.,  3.,
        3.,  0.,  3., nan, nan,  0., nan, nan,  0.,  0.,  0.,  0

In [42]:
hilton_merged['Cluster Labels']= hilton_merged['Cluster Labels'].replace(np.nan, (kclusters))

In [43]:
hilton_merged['Cluster Labels']= hilton_merged['Cluster Labels'].astype('int')

In [44]:
hilton_merged['Cluster Labels'].value_counts()

0    127
3     68
6     30
4      2
5      1
2      1
1      1
Name: Cluster Labels, dtype: int64

In [45]:
#check colors

x = np.arange(kclusters+1)
ys = [i + x + (i*x)**2 for i in range(kclusters+1)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]
rainbow

['#8000ff', '#2c7ef7', '#2adddd', '#80ffb4', '#d4dd80', '#ff7e41', '#ff0000']

In [46]:
# create map
map_clusters = folium.Map(width=800,height=800, location=[center_lat, center_long], zoom_start=4)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(hilton_merged['Latitude'], hilton_merged['Longitude'], hilton_merged['City'], hilton_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [130]:
# Reset the dataframe for more trials
hilton_venues_sorted = pd.DataFrame(columns=columns)
hilton_venues_sorted['Neighborhood'] = hilton_grouped['Neighborhood']

for ind in np.arange(hilton_grouped.shape[0]):
    hilton_venues_sorted.iloc[ind, 1:] = return_most_common_venues(hilton_grouped.iloc[ind, :], num_top_venues)


In [131]:
# set number of clusters
kclusters = 20

hilton_grouped_clustering = hilton_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, n_init=10, random_state=0).fit(hilton_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([ 4,  4,  2,  4,  4,  4,  4,  4,  4, 18], dtype=int32)

In [132]:
# add clustering labels
hilton_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

hilton_merged = df

# merge hilton_grouped with df to add latitude/longitude for each neighborhood
hilton_merged = hilton_merged.join(hilton_venues_sorted.set_index('Neighborhood'), on='City')

#check the impact
hilton_merged['Cluster Labels'].value_counts()

4.0     110
18.0     42
8.0      12
5.0       9
0.0       8
12.0      5
2.0       1
3.0       1
15.0      1
16.0      1
19.0      1
9.0       1
1.0       1
11.0      1
7.0       1
13.0      1
10.0      1
6.0       1
14.0      1
17.0      1
Name: Cluster Labels, dtype: int64

In [134]:
hilton_merged['Cluster Labels'].values

array([ 4.,  4.,  2., nan,  4., nan,  4.,  4.,  4.,  4.,  4., 18.,  4.,
        4., nan, 18.,  3., 12.,  4.,  4., 18.,  4.,  4.,  4.,  0.,  4.,
       18., 12., 15.,  4., 12.,  4., 18.,  4., 18.,  4.,  4.,  4.,  0.,
        4., 18., 18., nan, 18., 18., 18.,  4.,  4.,  4., 16.,  4., nan,
       18., 18., 18., 18.,  8., nan, 18., 12.,  4.,  4., 18.,  4.,  4.,
        4.,  4., 18.,  4.,  4., 17., nan, nan,  4.,  4.,  8.,  5.,  4.,
        4.,  4., 18.,  4.,  1.,  5.,  5.,  4.,  4., 11.,  8., 18.,  4.,
        4.,  0., 18., 18.,  4., nan, 18., nan,  4.,  4., nan,  4.,  7.,
        4.,  4., 13.,  4.,  4., 10.,  4., nan,  4., nan,  4.,  4., nan,
        4.,  4., 18.,  4., 18.,  4.,  4.,  4.,  6.,  0.,  0.,  4., 18.,
        4.,  4., nan,  4.,  4., nan,  4.,  8., 18.,  4.,  5., nan, nan,
        4.,  4., 14., 18., 18.,  4.,  4.,  8., 18., 18.,  4.,  4., nan,
        4., nan,  4.,  8., 18.,  4., nan,  4.,  4.,  4., 12.,  4.,  8.,
        4.,  4.,  9., nan, nan,  4., nan, nan,  4.,  4., 18.,  4